File: examples/example-jackson2/src/test/java/com/example/jackson2/JacksonRoundTripTest.java
package com.example.jackson2;

import static org.assertj.core.api.Assertions.assertThat;

import com.example.dto.Address;
import com.example.dto.Blob;
import com.example.dto.LineItem;
import com.example.dto.Manifest;
import com.example.dto.Money;
import com.example.dto.Order;
import com.example.dto.Tag;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.util.Base64;
import java.util.List;
import org.junit.jupiter.api.Test;

/**
 * Serializes each generated model with Jackson 2's {@code ObjectMapper}, checks the produced JSON,
 * deserializes it back, and confirms the round trip is a different instance with the same value.
 */
class JacksonRoundTripTest {

  private final ObjectMapper mapper = new ObjectMapper();

  @Test
  void simpleType() throws Exception {
    Tag tag = Tag.builder().id("T1").label("urgent").build();

    String json = mapper.writeValueAsString(tag);
    JsonNode node = mapper.readTree(json);
    assertThat(node.get("id").asText()).isEqualTo("T1");
    assertThat(node.get("label").asText()).isEqualTo("urgent");
    // "priority" defaults to 0 and was never set: the constructor already applied the default, so
    // it serializes like any other value.
    assertThat(node.get("priority").asInt()).isEqualTo(0);

    Tag back = mapper.readValue(json, Tag.class);
    assertThat(back).isNotSameAs(tag).isEqualTo(tag);
  }

  @Test
  void aDefaultedFieldFallsBackWhenTheJsonOmitsItEntirely() throws Exception {
    // No "priority" key at all in the source JSON - the @JsonCreator constructor parameter comes
    // in as null, same as an unset builder field, and falls back to the schema default.
    Tag tag = mapper.readValue("{\"id\":\"T2\"}", Tag.class);

    assertThat(tag.getPriority()).isEqualTo(0);
  }

  @Test
  void typeWithManyNonNullableParams() throws Exception {
    Money money = Money.builder().currency("USD").amount(19.99).scale(2).taxable(true).build();

    String json = mapper.writeValueAsString(money);
    JsonNode node = mapper.readTree(json);
    assertThat(node.get("currency").asText()).isEqualTo("USD");
    assertThat(node.get("amount").asDouble()).isEqualTo(19.99);
    assertThat(node.get("scale").asInt()).isEqualTo(2);
    assertThat(node.get("taxable").asBoolean()).isTrue();

    Money back = mapper.readValue(json, Money.class);
    assertThat(back).isNotSameAs(money).isEqualTo(money);
  }

  @Test
  void typeWithNestedObject() throws Exception {
    Address address = Address.builder().street("1 Main St").city("Springfield").build();
    Order order = Order.builder().id("O1").shippingAddress(address).build();

    String json = mapper.writeValueAsString(order);
    JsonNode node = mapper.readTree(json);
    assertThat(node.get("shippingAddress").get("city").asText()).isEqualTo("Springfield");

    Order back = mapper.readValue(json, Order.class);
    assertThat(back).isNotSameAs(order).isEqualTo(order);
    assertThat(back.getShippingAddress()).isNotSameAs(address).isEqualTo(address);
  }

  @Test
  void typeWithNestedListOfObjects() throws Exception {
    List<LineItem> items =
        List.of(
            LineItem.builder().sku("A1").quantity(2).build(),
            LineItem.builder().sku("B2").quantity(1).build());
    Manifest manifest = Manifest.builder().id("M1").items(items).build();

    String json = mapper.writeValueAsString(manifest);
    JsonNode node = mapper.readTree(json);
    assertThat(node.get("items")).hasSize(2);
    assertThat(node.get("items").get(1).get("sku").asText()).isEqualTo("B2");

    Manifest back = mapper.readValue(json, Manifest.class);
    assertThat(back).isNotSameAs(manifest).isEqualTo(manifest);
    assertThat(back.getItems()).isNotSameAs(items).isEqualTo(items);
  }

  @Test
  void aBytesFieldSerializesAsABase64StringAndRoundTrips() throws Exception {
    byte[] payload = {1, 2, 3, 4, 5};
    byte[] signature = "sig".getBytes();
    Blob blob = Blob.builder().id("B1").payload(payload).signature(signature).build();

    String json = mapper.writeValueAsString(blob);
    JsonNode node = mapper.readTree(json);
    // Jackson renders byte[] as a base64 string out of the box - no custom serializer needed.
    assertThat(node.get("payload").isTextual()).isTrue();
    assertThat(node.get("payload").asText()).isEqualTo(Base64.getEncoder().encodeToString(payload));
    assertThat(node.get("signature").asText())
        .isEqualTo(Base64.getEncoder().encodeToString(signature));

    Blob back = mapper.readValue(json, Blob.class);
    assertThat(back).isNotSameAs(blob).isEqualTo(blob);
    assertThat(back.getPayload()).isNotSameAs(payload).containsExactly(payload);

    // toString prints the same base64, not a byte-array dump
    assertThat(blob.toString())
        .contains("payload=" + Base64.getEncoder().encodeToString(payload))
        .contains("signature=" + Base64.getEncoder().encodeToString(signature));
  }

  @Test
  void bytesAccessorsAllReturnAndStoreDefensiveCopies() throws Exception {
    byte[] source = {10, 20, 30};
    Blob blob = Blob.builder().id("B2").payload(source).build();

    // mutating the array passed to the builder must not affect the built DTO
    source[0] = 99;
    assertThat(blob.getPayload()).containsExactly(10, 20, 30);

    // the getter hands back a fresh copy every time
    byte[] first = blob.getPayload();
    byte[] second = blob.getPayload();
    assertThat(first).isNotSameAs(second);
    first[0] = 42;
    assertThat(blob.getPayload()).containsExactly(10, 20, 30);
  }
}

File: examples/example-jackson/src/test/kotlin/com/example/jackson/JacksonRoundTripTest.kt
package com.example.jackson

import com.example.dto.Address
import com.example.dto.LineItem
import com.example.dto.Manifest
import com.example.dto.Money
import com.example.dto.Order
import com.example.dto.Shipment
import com.example.dto.Tag
import com.example.dto.mutate
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import tools.jackson.databind.json.JsonMapper
import tools.jackson.module.kotlin.KotlinModule

/**
 * Serializes each generated model with Jackson 3's `JsonMapper` (jackson-module-kotlin registered),
 * checks the produced JSON, deserializes it back, and confirms the round trip is a different
 * instance with the same value.
 */
class JacksonRoundTripTest {

  private val mapper = JsonMapper.builder().addModule(KotlinModule.Builder().build()).build()

  @Test
  fun `simple type`() {
    val tag = Tag.builder().id("T1").label("urgent").build()

    val json = mapper.writeValueAsString(tag)
    val node = mapper.readTree(json)
    assertThat(node.get("id").asString()).isEqualTo("T1")
    assertThat(node.get("label").asString()).isEqualTo("urgent")

    val back = mapper.readValue(json, Tag::class.java)
    assertThat(back).isNotSameAs(tag).isEqualTo(tag)

    // modelmaker.kotlin.enabled: the generated Kotlin `mutate { }` extension over the builder.
    val relabeled = back.mutate { label("stale") }
    assertThat(relabeled.label).isEqualTo("stale")
    assertThat(back.label).isEqualTo("urgent")
  }

  @Test
  fun `type with many non-nullable params`() {
    val money = Money.builder().currency("USD").amount(19.99).scale(2).taxable(true).build()

    val json = mapper.writeValueAsString(money)
    val node = mapper.readTree(json)
    assertThat(node.get("currency").asString()).isEqualTo("USD")
    assertThat(node.get("amount").asDouble()).isEqualTo(19.99)
    assertThat(node.get("scale").asInt()).isEqualTo(2)
    assertThat(node.get("taxable").asBoolean()).isTrue()

    val back = mapper.readValue(json, Money::class.java)
    assertThat(back).isNotSameAs(money).isEqualTo(money)
  }

  @Test
  fun `type with nested object`() {
    val address = Address.builder().street("1 Main St").city("Springfield").build()
    val order = Order.builder().id("O1").shippingAddress(address).build()

    val json = mapper.writeValueAsString(order)
    val node = mapper.readTree(json)
    assertThat(node.get("shippingAddress").get("city").asString()).isEqualTo("Springfield")

    val back = mapper.readValue(json, Order::class.java)
    assertThat(back).isNotSameAs(order).isEqualTo(order)
    assertThat(back.shippingAddress).isNotSameAs(address).isEqualTo(address)
  }

  @Test
  fun `type with nested list of objects`() {
    val items =
        listOf(
            LineItem.builder().sku("A1").quantity(2).build(),
            LineItem.builder().sku("B2").quantity(1).build(),
        )
    val manifest = Manifest.builder().id("M1").items(items).build()

    val json = mapper.writeValueAsString(manifest)
    val node = mapper.readTree(json)
    assertThat(node.get("items")).hasSize(2)
    assertThat(node.get("items").get(1).get("sku").asString()).isEqualTo("B2")

    val back = mapper.readValue(json, Manifest::class.java)
    assertThat(back).isNotSameAs(manifest).isEqualTo(manifest)
    assertThat(back.items).isNotSameAs(items).isEqualTo(items)
  }

  @Test
  fun `enum-as-pattern field and long or float number formats`() {
    val shipment =
        Shipment.builder()
            .id("S1")
            .status("IN_TRANSIT")
            .distance(9_000_000_000L)
            .weight(72.5f)
            .build()

    val json = mapper.writeValueAsString(shipment)
    val node = mapper.readTree(json)
    assertThat(node.get("status").asString()).isEqualTo("IN_TRANSIT")
    // distance exceeds Int.MAX_VALUE - only representable because "format": "long" renders Long.
    assertThat(node.get("distance").asLong()).isEqualTo(9_000_000_000L)
    assertThat(node.get("weight").asDouble()).isEqualTo(72.5)

    val back = mapper.readValue(json, Shipment::class.java)
    assertThat(back).isNotSameAs(shipment).isEqualTo(shipment)
  }

  @Test
  fun `an unset optional enum field is filled in with its default`() {
    val shipment = Shipment.builder().id("S1").status("NEW").distance(10L).weight(1.0f).build()

    // "priority" was never set - "default": "NORMAL" is applied by the constructor.
    assertThat(shipment.priority).isEqualTo("NORMAL")

    val json = mapper.writeValueAsString(shipment)
    assertThat(mapper.readTree(json).get("priority").asString()).isEqualTo("NORMAL")

    val urgent = shipment.mutate { priority("HIGH") }
    assertThat(urgent.priority).isEqualTo("HIGH")
    assertThat(shipment.priority).isEqualTo("NORMAL")
  }
}
